### ISSUE TITLE
Performance: fw_get_db_settings_option vs. get_option

### ISSUE THREAD

**Issue by gonzela2006** [ThemeFuse/Unyson#2709]
Hello,

I was making a performance check and found that `fw_get_db_settings_option` function is so slow compared with WordPress `get_option` function

```php
$t1 = microtime( true );
$option_id = fw_get_db_settings_option( 'option_id' );
$t2 = microtime( true );
print_r( $t2 - $t1 );
```
Load time is: `0.031062126159668`

```php
$t1 = microtime( true );
$option_id = get_option( 'fw_theme_settings_options:default' )['option_id'];
$t2 = microtime( true );
print_r( $t2 - $t1 );
```
Load time is: `0.00066804885864258`

`get_option` is %97.85 faster.

Why `fw_get_db_settings_option` is so slow?and How can we improve it?

Thanks

**Comment by andreiglingeanu** [comment]
The truth is `fw_get_db_settings_option` does a lot more than Core's `get_option`. It does options defaults extraction, which implies opening multiple `options.php` files. [Here's](https://github.com/ThemeFuse/Unyson/blob/master/framework/helpers/database.php#L64-L66) where that function is defined, you are welcome to try your hand at narrowing down the most hot place in it. Thanks!

**Comment by ViorelEremia** [comment]
You can use one of most popular cache plugins https://wordpress.org/plugins/search/cache/ or you can create own cache with default wordpress cache object https://codex.wordpress.org/Class_Reference/WP_Object_Cache
I wrote by fly a prototype function but i not test it be carrrefull how you use 
`function my_prefix_set( $key, $default = '' ) {

	global $my_prefix_set;

	if ( ! $my_prefix_set ) {

		if ( false === ( $my_prefix_set = get_transient( 'my_prefix_settings' ) ) ) {
			$my_prefix_set = fw_get_db_settings_option();
			set_transient( 'my_prefix_settings', $my_prefix_set, 300000 * HOUR_IN_SECONDS );
		}
	}

	/*
	 To extract key by prototype key1/key2/key3 you can do as in that function
	 https://github.com/ThemeFuse/Unyson/blob/master/framework/helpers/class-fw-db-options-model.php#L111
	*/
	return isset( $my_prefix_set[ $key ] ) ? $my_prefix_set[ $key ] : $default;
}`

You can use transient or wp_add_cache get_option etc
But I recommended to use un cache plugin, be careful don't use transient with w3total cache they have a bug :) may have solved from now I do not know.

**Comment by danyj** [comment]
@ViorelEremia as far as I can see this is a legit question about Unyson performance and improvement so we should  not include third party plugins but look for possible framework solutions if that would increase the `fw_get_db_settings_option` performance.

**Comment by andreiglingeanu** [comment]
@ViorelEremia To be fair, using a transient or a general purpose caching plugin won't quite work for improving `fw_get_db_*` functions. The reason is the following.

<hr>

**For transients**

When some admin will change an option from theme settings (or some modal), they'll have to wait for the _transient_ to expire, before seeing the change being reflected in frontend. That is kind of a bummer.

<hr>

**General purpose caching plugins**

The situation is even worse for caching plugins because, besides the timeout, they provide long-term caching. This long-term caches are then invalidated at common WordPress core hooks and actions (post save, post edit and so on).

This may lead to options being cached and never invalidated, which creates a lot of confusion!

<hr>

The bottom line is that it is hard to cache the result of `fw_get_db_*` functions, because the cache invalidation is very hard. This is especially hard at achieving at the theme level.

We had a discussion with @GheorgheP about introducing the idea of `OptionContext`, which will allow us to perform at least some sort of caching at the framework level.

**Comment by danyj** [comment]
> Save/Reset is slow and retrieval in frontend becomes slow too

amen!
